extensions/Worker.Extensions.Storage.Queues/src/TypeConverters/QueueMessageConverter.cs

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Azure.Storage.Queues.Models;
+using Microsoft.Azure.Functions.Worker.Converters;
+using Microsoft.Azure.Functions.Worker.Core;
+using Microsoft.Azure.Functions.Worker.Extensions.Abstractions;
+using Microsoft.Azure.Functions.Worker.Storage.Queues;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Functions.Worker
+{
+    [SupportsDeferredBinding]
+    [SupportedConverterType(typeof(QueueMessage))]
+    internal sealed class QueueMessageConverter : QueueConverterBase<QueueMessage>
+    {
+        private readonly JsonSerializerOptions _jsonOptions;
+
+        public QueueMessageConverter(ILogger<QueueMessageConverter> logger) : base(logger)
+        {
+            _jsonOptions = new() { Converters = { new QueueMessageJsonConverter() } };
+        }
+
+        public override ValueTask<ConversionResult> ConvertAsync(ConverterContext context)
+        {
+            if (!CanConvert(context))
+            {
+                return new ValueTask<ConversionResult>(ConversionResult.Unhandled());
+            }
+
+            try
+            {
+                var modelBindingData = (ModelBindingData)context.Source!;
+                QueueMessage queueMessage = ExtractQueueMessage(modelBindingData);
+                return new ValueTask<ConversionResult>(ConversionResult.Success(queueMessage));
+            }
+            catch (JsonException ex)
+            {
+                string msg = String.Format(CultureInfo.CurrentCulture,
```

Comment:
There is a fair bit of common logic between this `ConvertAsync` and `QueueMessageBinaryDataConverter`'s implementation. You could extract the common logic to the base class you already have, then add `protected abstract` methods for the specialized logic.

``` CSharp
abstract class QueueMessageConverterBase
{
    public ValueTask<ConversionResult> ConvertAsync(...)
    {
        // common logic
    }

    protected abstract ValueTask<T> ConvertCoreAsync(ModelBindingData data);
}
```

Reply:
great suggestion - thanks!